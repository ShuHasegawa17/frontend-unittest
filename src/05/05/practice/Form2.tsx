import { useId, useState } from 'react';
import { Agreement2 } from './Agreement2';
import { InputAccount2 } from './InputAccount2';

export const Form2 = () => {
  const [checked, setChecked] = useState(false);
  const headingId = useId();
  return (
    <form aria-labelledby={headingId}>
      <h2 id={headingId}>新規アカウント登録</h2>
      <InputAccount2 />
      <Agreement2
        onChange={(event) => {
          setChecked(event.currentTarget.checked);
        }}
      />
      <div>
        <button disabled={!checked}>サインアップ</button>
      </div>
    </form>
  );
};
