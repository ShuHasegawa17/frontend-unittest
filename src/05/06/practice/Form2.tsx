import { useState } from 'react';
import { ContactNumber } from '../ContactNumber';
import { DeliveryAddress } from '../DeliveryAddress';
import { PastDeliveryAddress } from '../PastDeliveryAddress';
import { RegisterDeliveryAddress } from '../RegisterDeliveryAddress';

// & は交差型　2つの型を統合 = interface の extendsと同じ
export type AddressOption = React.ComponentProps<'option'> & { id: string };
export type Props = {
  deliveryAddresses?: AddressOption[];
  onSubmit?: (event: React.FormEvent<HTMLFormElement>) => void;
};

export const Form2 = (props: Props) => {
  const [registerNew, setRegisterNew] = useState<boolean | undefined>(
    undefined
  );
  return (
    <form onSubmit={props.onSubmit}>
      <h2>お届け先情報の入力</h2>
      <ContactNumber />
      {props.deliveryAddresses?.length ? (
        <>
          <RegisterDeliveryAddress onChange={setRegisterNew} />
          {registerNew ? (
            <DeliveryAddress title="新しいお届け先" />
          ) : (
            <PastDeliveryAddress
              disabled={registerNew === undefined}
              options={props.deliveryAddresses}
            />
          )}
        </>
      ) : (
        <DeliveryAddress />
      )}
      <hr />
      <div>
        <button>注文内容の確認へ進む</button>
      </div>
    </form>
  );
};
