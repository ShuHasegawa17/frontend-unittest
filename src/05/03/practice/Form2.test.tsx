import { fireEvent, logRoles, render, screen } from '@testing-library/react';
import { Form2 } from './Form2';

describe('UIコンポーネントテスト', () => {
  test('名前の表示', () => {
    // レンダリング
    render(<Form2 name="hase" />);
    // テキストを取得
    const text = screen.getByText('hase');
    // 取得したテキストがDOMに存在するか。
    expect(text).toBeInTheDocument();
  });
  test('ボタンの表示', () => {
    render(
      <>
        <Form2 name="chisa" />
      </>
    );
    expect(screen.getByRole('button')).toBeInTheDocument();
  });
  test('見出しの表示', () => {
    render(<Form2 name="hase" />);
    expect(screen.getByRole('heading')).toHaveTextContent('アカウント情報');
  });
  test('ボタンを押すとイベントハンドラーが実行される', () => {
    // モック関数をサブミット引数に渡す。
    const mockFn = jest.fn();
    render(<Form2 name="hase" onSubmit={mockFn} />);
    // クリック
    fireEvent.click(screen.getByRole('button'));
    // モック関数が呼ばれたこと。
    expect(mockFn).toHaveBeenCalled();
  });
});
test('snap', () => {
  const { container } = render(<Form2 name="hase" />);
  expect(container).toMatchSnapshot();
});
test('logRoles: レンダリング結果からロールとアクセシブルネームを確認', () => {
  const { container } = render(<Form2 name="taro" />);
  logRoles(container);
});
