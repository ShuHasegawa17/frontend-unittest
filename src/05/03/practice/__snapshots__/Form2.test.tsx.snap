// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snap 1`] = `
<div>
  <form>
    <h2>
      アカウント情報
    </h2>
    <p>
      hase
    </p>
    <div>
      <button>
        編集する
      </button>
    </div>
  </form>
</div>
`;
